src/main/jwks.ts

```diff
@@ -0,0 +1,120 @@
+import Ajv from 'ajv';
+import { Request, FetchOptions, Response } from '@automationcloud/request';
+import { Exception } from './exception';
+import { ajvErrorToMessage } from './util';
+
+const ajv = new Ajv({
+    allErrors: true,
+    useDefaults: true,
+    jsonPointers: true,
+    format: 'full',
+});
+
+// eslint-disable-next-line import/no-commonjs
+const validateFunction = ajv.compile(require('../../schema/ac-jwks.json'));
+
+export class JwksClient {
+    protected _cache: Map<string, JwksCache> = new Map();
+    request: Request;
+
+    constructor(protected options: JwksOptions) {
+        this.options = options;
+        this.request = new Request({
+            baseUrl: this.options.url,
+            retryAttempts: this.options.retryAttempts ?? 3,
+            fetch: this.options.fetch,
+        });
+    }
+
+    async getSigningKey(): Promise<string> {
+        const alg = this.options.algorithm;
+        const cached = this.cache.get(alg);
+        if (cached) {
+            return cached;
+        }
+
+        const res = await this.request.get('');
+        const { keys } = this.validateResponse(res);
+        const matchingKey = keys.find(k => k.alg === this.options.algorithm);
+        if (!matchingKey) {
+            throw new Exception({
+                name: 'SigningKeyNotFoundError',
+                message: 'Expected signing key not found from response',
+            });
+        }
+
+        this.cache.set(matchingKey.alg, matchingKey.k);
+
+        return matchingKey.k;
+    }
+
+    get cache() {
```

Comment:
Interestingly, I think this getter will add more GC load if called too frequently, since it implicitly creates an instance per call. Wondering if we could just store the entire key payload (not per-alg) and thus avoid this. Also feel free to not do anything, since the impact might be negligible in grand scheme (e.g. if it's just 1 call per request, it's fine)

Reply:
Thanks, addressed it 🙏